modules/syslogformat/syslog-format.c

```diff
@@ -336,10 +354,24 @@ _syslog_format_parse_legacy_program_name(LogMessage *msg, const guchar **data, g
   src = *data;
   left = *length;
   prog_start = src;
+  gboolean has_alpha_char = FALSE;
+
   while (left && *src != ' ' && *src != '[' && *src != ':')
     {
+      if (G_UNLIKELY(flags & LP_CHECK_PROGRAM) && !_validate_program_char(*src, &has_alpha_char))
+        {
+          log_msg_set_tag_by_id(msg, LM_T_SYSLOG_RFC_3164_INVALID_PROGRAM);
+          return;
+        }
       _skip_char(&src, &left);
     }
+
+  if (G_UNLIKELY(flags & LP_CHECK_PROGRAM) && !has_alpha_char)
+    {
+      log_msg_set_tag_by_id(msg, LM_T_SYSLOG_RFC_3164_INVALID_PROGRAM);
+      return;
```

Comment:
We can't early return from here, as we need to optionally set the legacy message header:

```
if ((flags & LP_STORE_LEGACY_MSGHDR))
...
```

(In case of an invalid program name, the header should not contain the program name)